design/src/rfcs/rfc0036_http_dep_elimination.md

```diff
@@ -0,0 +1,107 @@
+
+<!-- Explain the implementation of this new feature -->
+How to actually implement this RFC
+----------------------------------
+
+### Enabling API evolution
+One key mechanism that we SHOULD use for allowing our APIs to evolve in the future is usage of `~` version bounds for the runtime crates after releasing 1.0.
+
+### Http Request Wrapper
+
+In order to enable HTTP evolution, we will create a set of wrapper structures around `http::Request` and `http::Response`. These will use `http = 0` internally. Since the HTTP crate itself is quite small, including private dependencies on both versions of the crate is a workable solution. In general, we will aim for an API that is close to drop-in compatible to the HTTP crate while ensuring that a different crate could be used as the backing storage.
+
+```rust
+// since it's our type, we can default `SdkBody`
+pub struct Request<B = SdkBody> {
+    // this uses the http = 0.2 request. In the future, we can make an internal enum to allow storing an http = 1
+    http_0: http::Request<B>
+}
+```
+
+**Conversion to/from `http::Request`**
+One key property here is that although converting to/from an `http::Request` **can** be expensive, this is *not* ossification of the API. This is because the API can support converting from/to both `http = 0` and `http = 1` in the future—because it offers mutation of the request via a unified interface, the request would only need to be converted once for dispatch if there was a mismatch (instead of repeatedly). At some point in the future, the `http = 0` representation could be deprecated and removed or feature gated.
+
+**Challenges**
+1. Creating an HTTP API which is forwards compatible, idiomatic and "truthful" without relying on existing types from Hyper—e.g. when adding a header, we need to account for the possibility that a header is invalid.
+2. Allow for future forwards-compatible evolution in the API—A lot of thought went into the `http` crate API w.r.t method parameters, types, and generics. Although we can aim for a simpler solution in some cases (e.g. accepting `&str` instead of `HeaderName`), we need to be careful that we do so while allowing API evolution.
+
+### Removing the SigV4 HTTP dependency
+The SigV4 crate signs a number of `HTTP` types directly. We should change it to accept strings, and when appropriate, iterators of strings for headers.
+
+### Removing the HTTP dependency from generated clients
+Generated clients currently include a public HTTP dependency in `customize`. This should be changed to accept our `HTTP` wrapper type instead or be restricted to a subset of operations (e.g. `add_header`) while forcing users to add an interceptor if they need full control.
+
+
+In order to implement this feature, we need to add X and update Y...
+
+<!-- Include a checklist of all the things that need to happen for this RFC's implementation to be considered complete -->
+Changes checklist
+-----------------
```

Comment:
We also want an item to remove `http` types from `external-types.xml`s?

Reply:
we actually can't do that—we _may_ want the ability to have something like `#[allow_external]` somehow for the external types checker

Reply:
To clarify, the only allowed instances are the ones that are used exclusively from generated code?

Reply:
no, the wrapper type itself will have conversion functions so that e.g. we can use it with Hyper

Reply:
but it should be removed from almost all locations